[Ready] Adds Support For Player Controlled Megafauna + Megafauna Code Refactors

Whoneedspacee opened the pull request:
## About The Pull Request

Fireball code has been moved back to the drake file. Speed of megafauna has been set to the move_to_delay so players have the same actual speed as the megafauna. Megafauna now have action abilities for players controlling them to set which ability they want to use. Also refactors a bunch of the megafauna code to have more consistent places in files. Removes the stupid code in blood drunk miner that MSO commented on and removes a useless check in the legion megafauna file.

Also has some minor changes. Bubblegum blood cannot track anymore due to it causing lag during fights with thousands of bloody footsteps stacked on eachother. Adds some new icons for the unique blood step as well. Hierophant controlled players can no longer alt click to teleport instantly to any space. Lesser ash drake is also the only one with an alt click swoop now as the main drake has no use for it anymore. Also gives space dragon an action button for its tail swoop spell (I seriously had no idea it even existed why didn't it have this in the first place). The delay after the lava arena attack is also slightly increased in order to give more time for shots to go off as well.

## Why It's Good For The Game

Generally megafauna aren't player controlled obviously, but events might have them as well as maybe some meme stuff. This gives them some choice in what they want to do instead of it literally just being random. Trying to just attack one person? Too bad you started the attack that destroys the entire station.

Also fixes up some frankly stupid code stuff that was probably just laziness at the time.

Here's a display of what it looks like: [Player Controlled Megafauna](https://www.youtube.com/watch?v=IIJ68ZVPVrw&feature=youtu.be)

## Changelog
:cl: Joe Berry
add: Player controlled megafauna support. Now if you control any megafauna, or the space dragon, you have action buttons to choose the different moves you can use.
refactor: Fixed up megafauna code.
/:cl:

Whoneedspacee committed:
adds small sprites for many of the large megafauna

moves fireball target code back to drake file

changes the speed of every megafauna mob to its move to delay so player controlled ones have the same speed

adds action abilities to each megafauna so players can select the type of attack they want

refactors pretty much all of the megafauna code to be simpler to edit and easier to read as well as have some consistency in where stuff like initialize goes

lesser ash drake now inherits alt click on instead of normal ash drake having it

hierophant no longer blinks on alt click as well

removes pointless check in legion code for ranged timer delay as the proc that calls openfire already checks that

adds icon for new bubblegum blood tracks

bubblegum blood no longer leaves footstep trails as it caused massive lag when many blood tiles were run over in a fight as well as being way too excessive

optimumtact commented:
what a stupid stupid idea

Tlaltecuhtli commented:
time to tc trade for bubblegum

Barhandar commented:
>Speed of megafauna has been set to the move_to_delay so players have the same actual speed as the megafauna.

As in, players possessing megafauna have same speed as the AI-controlled version, or as in you're fucked and can't run away from megafauna anymore?
>Bubblegum blood cannot track anymore due to it causing lag during fights with thousands of bloody footsteps stacked on eachother.  
>>hurr durr lets remove main danger of bubbley fights because the blood-tracking code can't squash footsteps

Also, spawning megafauna on station is grounds for immediate deadmining. Spawning PLAYER CONTROLLED megafauna is worse, even space dragon with its 1/10th HP of the real ash drake and IIRC rediced damage can _easily_ obliterate everything because there's a player driving it. The only occasion possessed megafauna should happen is admeme events on dedicated arenas, where your concerns about being picky in attacks do not matter.

Whoneedspacee commented:
> > Speed of megafauna has been set to the move_to_delay so players have the same actual speed as the megafauna.
> 
> As in, players possessing megafauna have same speed as the AI-controlled version, or as in you're fucked and can't run away from megafauna anymore?
> 
> > Bubblegum blood cannot track anymore due to it causing lag during fights with thousands of bloody footsteps stacked on eachother.
> > > hurr durr lets remove main danger of bubbley fights because the blood-tracking code can't squash footsteps
> 
> Also, spawning megafauna on station is grounds for immediate deadmining. Spawning PLAYER CONTROLLED megafauna is worse, even space dragon with its 1/10th HP of the real ash drake and IIRC rediced damage can _easily_ obliterate everything because there's a player driving it. The only occasion possessed megafauna should happen is admeme events on dedicated arenas, where your concerns about being picky in attacks do not matter.

Players possessing have the same speed as the AI version. So you aren't fucked when the colossus can move 3 times as fast as the AI version.

The bubblegum fight will be fine, hallucinations leave blood trails and so does bubblegum himself. There will be more than enough blood to fuck people over, it just won't stack on itself 5 times over. I'm the one who designed the fight in the first fucking place give me some credit man.

You're retarded if you think the space dragon can take down the entire station first of all. If that were true then people would only kill the ash drake for lesser ash drake since they're pretty much the same thing. But they don't because it's not that much better than someone robust with just really good healing shit and a spellblade or something more consistent to get.

/tg/station is not the only server, other servers could be fine summoning megafauna for events, or something really weak like blood drunk miner as a thunderdome type thing. I'm not assuming it's going to be spawned on station, but I've seen it spawned even on /tg/station for very special events.

I see no reason to at least not add something like this.

Barhandar commented:
>You're retarded if you think the space dragon can take down the entire station first of all. 

I have _personally witnessed_ two different space dragons taking down the entire station, you fucking serverbanned downstreamer.

Whoneedspacee commented:
> > You're retarded if you think the space dragon can take down the entire station first of all.
> 
> I have _personally witnessed_ two different space dragons taking down the entire station, you fucking serverbanned downstreamer.

I'm not serverbanned anymore. It's sad that two space dragons took down an entire station when one person with a riot shotgun can kill a space dragon in a few shots. It's not strong.

Barhandar commented:
>>one person with incredibly specific piece of equipment that isn't a popular powergaming target can kill someone who follows the concept of "spacing" and hence doesn't even get into said specific piece of equipment's range

Daxxed commented:
Good quality of life updates, no buffs but it will defenitely help miners.
Can't wait to be a controlled bubblegum, has been a dream ever since you buffed it.
Sadly I don't think any sane admin would allow it.
Maybe if you make the bet to summon it to station and then IF you survive it you could trade all your TC and the next 3 antag rolls for it ( I am kidding it would still be disgusting and cause deadminning )

4dplanner commented:
Barhandarposting reaches a peak

You can already control megafauna, it's not like it being broken is why admins don't use it. 

Also blood tracking hasn't been the biggest threat for a while LOL

4dplanner commented:
If it was, people would just take light step/no shoes like the old days anyway

Shadowflame909 commented:
You won't be able to tc trade a traitor or nukie for it

But I've traded a wizard to be a sentient megafauna.

Time for a smoother ride.

anconfuzedrock commented:
Megafauna are undeniably miles stronger than a wizard. Yes, 2 combat shotguns will seriously dent their health, so it’s a shame the game isn’t balanced around combat shotguns. It’s balanced around disablers and sometimes lasers that they take half damage from. Even if we give the station all revolvers, we run into this problem where megafauna fucking destroy everything. The space dragon, a much weaker drake, has proven to maul an escape shuttle easily, but one on station for 2 minutes will force a shuttle call since it smashes all the rwalls to space. One burst of a colossus will ruin a station. Don’t you fucking DARE say they’re fair.

Shadowflame909 commented:
The detectives shotgun can brig bubblegum from 100 to 60.

Well Lexia black was the detective that round.

Point is, all megafaunas have burn resistance. So your wasting time with pulse rifles and lasers. If you have bullets you win

ExcessiveUseOfCobblestone commented:
> shotguns aren't popular powergaming targets

and they say coders don't play

NoxVS commented:
They usually aren’t too popular with how bulky they are to carry and their rarity. There are I think 4 that exist at round start and 3 are pretty much unobtainable by normal crew members.

4dplanner commented:
All convincing complaints so far revolve around admins misusing the features. I think it should just be clearly communicated that a sentient bubblegum most likely *will* wipe the station if at all competent, and they should be saved for round-enders or some other special case. That shouldn't be a reason to reject this PR, since we already require a lot of faith in admins to not screw up the round.

4dplanner commented:
Basically deadmin anyone who lets a wizard trade for megafauna before OR after this PR

Militaires commented:
>shotguns aren't popular powergaming targets

Barhandar reaches his zenith....

Whoneedspacee commented:
i don’t even know how you guys got to talking about the space dragon.

i added ONE ACTION BUTTON for an already added spell. it already had tail swipe and you could use it in a panel this just adds an action button.

that’s it. the only change to space dragon. epic gamer.

this pr is support for events that will basically never happen, will only happen when its meme destroy everything mode, but will indeed happen.

JJRcop commented:
Instead of action buttons to switch modes, capture right clicks and pop open a radial menu to switch.

Whoneedspacee commented:
but why, why would i do that

drderp3635 commented:
muh balance implications for cleaning up code, and implementing better controls for admin memes.

God you monster. This is just like that nightmare of a button that spawns a deathsquad, they are just too strong! Nobody in the current combat meta could take down an entire deathsquad reliably on green alert! I bet you if an admin pressed the deathsqaud button for no reason they would get in trouble!

Tlaltecuhtli commented:
radial menus dont fit, the various attacks should wprk like martial arts combos

MimicFaux commented:
> You're retarded if you think the space dragon can take down the entire station first of all.

For the 10,000 round event, I was a full fledged Ash Drake, and I was able to curb stomp all the specialized teams that were sent out to kill me. I was untouchable to the point I was able to sit in the bridge and make demands of the crew to bring me shit and start a treasure hoard.

restricting mega-fauna movement to mirror their AI counterparts is fine by me, but make sure that the player versions get the speed bonus as they approach death like the AI fauna get.

Whoneedspacee committed:
Merge branch 'master' into blooddrunkminer

Whoneedspacee committed:
fixes a megafauna bug where they could not attack while in space

megafauna can now move through space normally as they would anywhere else

bubblegum no longer runs from corpses and people dying

Whoneedspacee committed:
Merge branch 'blooddrunkminer' of https://github.com/Whoneedspacee/tgstation into blooddrunkminer

Whoneedspacee committed:
fix the weird line errors

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:
```diff
@@ -51,18 +51,55 @@ Difficulty: Medium
+			shoot_ka()
+		else if(chosen_attack == 3)
+			transform_weapon()
+		return
```
switch this

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -63,71 +62,81 @@ Difficulty: Medium
+			mass_fire()
+		else if(chosen_attack == 4)
 			lava_swoop()
+		return
```
switch

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -375,15 +389,40 @@ Difficulty: Medium
+	. = ..()
+
+/mob/living/simple_animal/hostile/megafauna/dragon/ex_act(severity, target)
+	if(severity == 3)
```
is this not defined?

Whoneedspacee replied:
it was just weird git stuff, i removed the dragon alt click and moved it to lesser dragon since it wasnt the actual dragon swoop it was a snowflaked version.

ExcessiveUseOfCobblestone replied:
when I say defined I mean the severity numbers like an `EXPLOSION_HEAVY` or something.

doesn't look like one exists tho so *shrug*

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm:
```diff
@@ -50,6 +49,56 @@ Difficulty: Medium
+	minimum_distance = 0
+	speed = 0
+	charging = 1
+	addtimer(CALLBACK(src, .proc/reset_charge), 50)
```
some of these look like boos

Whoneedspacee replied:
i didnt make the code for these i just moved them to their own procs so i’m not really sure what you’re referring to

all i can see is charging should be true

ExcessiveUseOfCobblestone replied:
yep

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm:
```diff
@@ -149,3 +159,21 @@
+	icon_icon = 'icons/mob/actions/actions_animal.dmi'
+	button_icon_state = ""
+	var/mob/living/simple_animal/hostile/megafauna/M
+	var/chosen_message = ""
```
can this be null?

Whoneedspacee replied:
besides the vars and icon_icon dont think i need button icon or name

ExcessiveUseOfCobblestone replied:
sorry, referring to `var/chosen_message = ""`

ExcessiveUseOfCobblestone reviewed (commented):
didn't look at hiero or bubbles yet

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -123,59 +75,78 @@ Difficulty: Hard
+		else if(chosen_attack == 2)
+			hallucination_charge()
+		else if(chosen_attack == 3)
+			surround_with_hallucinations()
```
switch

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -123,59 +75,78 @@ Difficulty: Hard
-		..()
+/mob/living/simple_animal/hostile/megafauna/bubblegum/proc/triple_charge()
+	charge(delay = 6)
+	charge(delay = 4) // The FitnessGram Pacer Test is a multistage aerobic capacity test that progressively gets more difficult as it continues.
```
mfw sued by fitnessgram for referencing their pacer test in an atmos sim

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -123,59 +75,78 @@ Difficulty: Hard
-		DestroySurroundings()
-	..()
+/mob/living/simple_animal/hostile/megafauna/bubblegum/proc/hallucination_charge()
+	if(health > maxHealth * 0.5)
```
if bubblegum's mechanic is he's empowered after an HP threshold it would be nice if that was defined so if I wanted to change it to say just 30% of his HP I could easily do so

`#define BUBBLEGUM_SMASH (health > maxhealth*0.5)` or something

ExcessiveUseOfCobblestone replied:
if each attack has a different threshold that just happens to be the same value in some cases then this can be ignored ofc

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -421,6 +367,141 @@ Difficulty: Hard
+	if(is_enraged())
+		visible_message("<span class='danger'>[src] deflects the projectile; [p_they()] can't be hit with ranged weapons while enraged!</span>", "<span class='userdanger'>You deflect the projectile!</span>")
+		playsound(src, pick('sound/weapons/bulletflyby.ogg', 'sound/weapons/bulletflyby2.ogg', 'sound/weapons/bulletflyby3.ogg'), 300, 1)
+		return 0
```
```suggestion
		return BULLET_ACT_BLOCK
```

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -421,6 +367,141 @@ Difficulty: Hard
+
+/mob/living/simple_animal/hostile/megafauna/bubblegum/CanPass(atom/movable/mover, turf/target)
+	if(istype(mover, /mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination))
+		return 1
```
sin

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/hierophant.dm:
```diff
@@ -83,135 +87,73 @@ Difficulty: Hard
+			blink_spam(blink_counter, target_slowness, cross_counter)
+		return
+
+	if(L && !blinking && L.stat == DEAD && get_dist(src, L) > 2)
```
```suggestion
	if(L?.stat == DEAD && !blinking && get_dist(src, L) > 2)
```

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/hierophant.dm:
```diff
@@ -440,10 +393,114 @@ Difficulty: Hard
+
+/mob/living/simple_animal/hostile/megafauna/hierophant/adjustHealth(amount, updating_health = TRUE, forced = FALSE)
+	. = ..()
+	if(src && . > 0 && !blinking)
```
does adjusthealth return the value healed/damaged? If it's just true...

```suggestion
	if(src && . && !blinking)
```

ExcessiveUseOfCobblestone replied:
also is the src check suppose to be `!QDELETED(src)`?

Whoneedspacee replied:
hierophant code is a mess to be honest and i don't really want to touch it all that much because i'm afraid i'll totally fuck it up but i'll try that

ExcessiveUseOfCobblestone replied:
lol me in every file, especially in code i've written previously

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/hierophant.dm:
```diff
@@ -50,7 +50,7 @@ Difficulty: Hard
 	armour_penetration = 50
```
this file has a lot of sleeps w/o sanity (do we just call those nightmares?)

Whoneedspacee replied:
4d was doing #define for sleeps that check for qdeleted or dead at the end of the sleep should probably replace every sleep in mobs with that type of stuff not gonna lie

ExcessiveUseOfCobblestone reviewed (commented):
ok I did the other 2 files

Whoneedspacee committed:
lots of sins have been cleansed (0 = false 1 = true)

switch cases for player attack selection

bubblegum has been bug fixed a bit (steps can be blood warped to, adds blood warp ability)

bubblegum now has some slight additions to its attacks to keep the fight faster paced

bubblegums enrage mode now lasts longer and makes him move faster

legion now has their speed changed properly when they charge (wow this is an old fucking bug like old old)

bubblegum now tries to be aggressive towards people that are laying down (no more resting under bubblegum charges without consequences)

bubblegum now spawns slaughterlings when he devours targets

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into blooddrunkminer

# Conflicts:
#	code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm
#	code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm

Whoneedspacee committed:
fine tunes new bubblegum attacks and replaces some is_procs with defines

Whoneedspacee committed:
fixes stupid snowflake stuff in bdm code

specifies in defines for bubblegum

lots of sleep_check_death now to avoid stupid stuff

removes need for hierophant_dying var as well as combines all the blast procs into one since they were so similar

kriskog commented:
Normal ash drakes now do the fire rain move, which they didnt before. Was this intentional? Its not in the changelog.

Whoneedspacee commented:
i put it in the about but not the changelog my bad.

it was a nerf though since he’d create lava below you beforehand.

kriskog commented:
It wasnt in the about either which is why I asked. It's fine though, just checking.

Whoneedspacee commented:
>Fireball code has been moved back to the drake file.

yeah my bad that was a really retarded way of saying that

kriskog commented:
You also readded slaughterlings to bubblegum? Why in the world

ExcessiveUseOfCobblestone commented:
the new tlat?

Whoneedspacee commented:
it was just this one pr papi please forgive me what a fucking mess

kriskog commented:
Why slaughterlings though? Did you look at new bubble and think to yourself: something that blocks your path while you are dodigng and your shots in the small windows of hitting him, thats just what he needs? Having 9 slaugherlings spawning every so often just makes the fight even more drawn out than it already was, and unless you got stormproof (which isn't guaranteed every round) you are probably just SOL on trying bubble.

Whoneedspacee commented:
he can only summon 3-5 depending on his anger modifier (damage taken)

he only summons them in phase 1

3 of them also spawn if he devours a corpse

Whoneedspacee commented:
if you want to discuss new attacks you want added to bubblegum instead i'd be very happy to listen though, this pr was just meant to address minor issues i'd seen like not enough varied attacks in bubblegum and i didn't really want to code entirely new attacks that much.
